### PR TITLE
[AOTI] Support .item() in the ABI-compatible mode

### PR DESCRIPTION
**Pull request by desertfire**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118024
* __->__ #117989

Summary:

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D52965076](https://our.internmc.facebook.com/intern/diff/D52965076)

### CODE REVIEW

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52965076).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52965076).

**Comment by desertfire** [comment]
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52965076).

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -2226,16 +2229,22 @@ def codegen_shape_tuple(self, shape: Tuple[Expr, ...]) -> str:
...
+            dtype_str = str(dtype).split(".")[-1]
+            self.writeline(f"{DTYPE_TO_CPP[dtype]} {node.sym};")
+            self.writeline(f"aoti_torch_item_{dtype_str}({data}, &{node.sym});")
         else:
-            convert_type = DTYPE_TO_ATEN[node.inputs[0].get_dtype()].replace(
-                "at::k", "to"
-            )
-            self.writeline(f"auto {node.sym} = {data}.item().{convert_type}();")
+            if node.is_bool:
+                self.writeline(f"bool {node.sym} = {data}.item() ? 1 : 0;")
```
Comment (ezyang) [inline torch/_inductor/codegen/wrapper.py:2242]: What's going on here lol

File: torch/_inductor/codegen/wrapper.py [hunk 2]
```diff
@@ -1861,6 +1861,9 @@ def generate_return(self, output_refs):
             arr_iface = config.use_minimal_arrayref_interface  # For brevity.
 
             def use_thread_local_cached_output_tensor(idx, output):
+                if self.cuda:
+                    return
```
Comment (ezyang) [inline torch/_inductor/codegen/wrapper.py:1865]: Just for my personal understanding, what is this doing?

**Review by ezyang — APPROVED** [review]
Looks legit. Is the testing not easy to do?

**Review by chenyang78 — APPROVED** [review]
LGTM. Thanks!

File: torch/_inductor/codegen/wrapper.py (hunk 2 above)
Comment (desertfire) [reply torch/_inductor/codegen/wrapper.py:1865]: `use_thread_local_cached_output_tensor` is an optimization the CPU AOTI uses, to make sure output tensors can reuse CPU memory.

File: torch/_inductor/codegen/wrapper.py (hunk 1 above)
Comment (desertfire) [reply torch/_inductor/codegen/wrapper.py:2242]: I don't really remember why this, but the `not config.aot_inductor.abi_compatible` branch will go away anyway.

**Comment by desertfire** [comment]
> Looks legit. Is the testing not easy to do?

We probably need the next scalar_to_tensor PR to do a proper testing for this one.

**Pull request closed without merging**